lib/frontend/src/Obelisk/Frontend.hs

```diff
@@ -0,0 +1,155 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Obelisk.Frontend
+  ( ObeliskWidget
+  , Frontend (..)
+  , runFrontend
+  ) where
+
+import Prelude hiding ((.))
+
+import Control.Category
+import Control.Concurrent
+import Control.Lens
+import Control.Monad
+import Control.Monad.Fix
+import Control.Monad.IO.Class
+import Control.Monad.Primitive
+import Control.Monad.Ref
+import Control.Monad.Trans.Class
+import Data.Dependent.Sum (DSum (..))
+import Data.Functor.Sum
+import Data.IORef
+import Data.Monoid hiding (Sum)
+import Data.Text (Text)
+import GHCJS.DOM hiding (bracket, catch)
+import GHCJS.DOM.Document
+import GHCJS.DOM.Node
+import qualified GHCJS.DOM.Types as DOM
+import Language.Javascript.JSaddle (JSM)
+import Obelisk.Route.Frontend
+import Reflex.Dom.Core
+import Reflex.Host.Class
+import qualified Reflex.TriggerEvent.Base as TriggerEvent
+
+makePrisms ''Sum
+
+type ObeliskWidget t x route m =
+  ( DomBuilder t m
+  , MonadFix m
+  , MonadHold t m
+  , MonadSample t (Performable m)
+  , MonadReflexCreateTrigger t m
+  , PostBuild t m
+  , PerformEvent t m
+  , MonadIO m
+  , MonadIO (Performable m)
+  , TriggerEvent t m
+  , HasDocument m
+  , MonadRef m
+  , Ref m ~ Ref IO
+  , MonadRef (Performable m)
+  , Ref (Performable m) ~ Ref IO
+  , MonadFix (Performable m)
+  , PrimMonad m
+  , Prerender x m
+  , EventWriter t (Endo route) m
+  )
+
+data Frontend route = Frontend
+  { _frontend_head :: !(forall t m x. ObeliskWidget t x route m => RoutedT t route m ())
+  , _frontend_body :: !(forall t m x. ObeliskWidget t x route m => RoutedT t route m ())
+  , _frontend_title :: !(route -> Text)
```

Comment:
Does this not introduce a restriction on the page title being determined **solely** from the route? 

In my particular case I fetch content from the backend (and this fetching happens inside `_frontend_body`), which returns the title (think markdown rendered page with a title extracted out of it), but I cannot use it in `_frontend_title`.

Reply:
Actually changing `_frontend_title` has no effect (at least on `ob run`); bug?

Reply:
Yep, it sure does!  Let's figure out what it really should be.

The initial reason I put this in here was to be able to give something to history.pushState; however, maybe there's another way we can achieve that without making this so inflexible.